src/bin/pg_autoctl/keeper.c

```diff
@@ -1333,3 +1318,198 @@ keeper_state_as_json(Keeper *keeper, char *json, int size)
+									   PG_AUTOCTL_REPLICA_USERNAME,
+									   authMethod))
+	{
+		log_error("Failed to edit HBA file \"%s\" to update rules to current "
+				  "list of nodes registered on the monitor",
+				  hbaFilePath);
+		return false;
+	}
+
+	/*
+	 * Only reload if Postgres is known to be running. If it's not running, we
+	 * edited the HBA and it's going to take effect at next restart of
+	 * Postgres, so we're good here.
+	 */
+	if (pg_setup_is_running(postgresSetup))
+	{
+		if (!pgsql_reload_conf(pgsql))
+		{
+			log_error("Failed to reload the postgres configuration after adding "
+					  "the standby user to pg_hba");
+			return false;
+		}
+	}
+
+	return true;
+}
+
+
+/*
+ * keeper_refresh_other_nodes call pgautofailover.get_other_nodes on the
+ * monitor and refreshes the keeper otherNodes array with fresh information,
+ * including each node's LSN position.
+ *
+ * When forceCacheInvalidation is true, instead of trusting our previous value
+ * for the keeper otherNodes array, keeper_refresh_other_nodes() instead runs
+ * through the whole monitor.get_other_nodes() result and updates HBA rules for
+ * all entries there. That's necessary after a pg_basebackup for instance.
+ * which will copy over the origin's pg_hba.conf.
+ */
+bool
+keeper_refresh_other_nodes(Keeper *keeper, bool forceCacheInvalidation)
+{
+	Monitor *monitor = &(keeper->monitor);
+	KeeperConfig *config = &(keeper->config);
+	PostgresSetup *postgresSetup = &(keeper->postgres.postgresSetup);
+
+	NodeAddressArray *otherNodesArray = &(keeper->otherNodes);
+	NodeAddressArray oldNodesArray = *otherNodesArray;
+	NodeAddressArray diffNodesArray = { 0 };
+
+	char *host = config->nodename;
+	int port = postgresSetup->pgport;
+
+	log_trace("keeper_refresh_other_nodes");
+
+	if (!monitor_get_other_nodes(monitor, host, port, ANY_STATE, otherNodesArray))
+	{
+		log_error("Failed to get_other_nodes() on the monitor");
+		return false;
+	}
+
+	/* compute nodes that need an HBA change (new ones, new hostnames) */
+	if (forceCacheInvalidation)
+	{
+		diffNodesArray = *otherNodesArray;
+	}
+	else
+	{
+		(void) diff_nodesArray(&oldNodesArray, otherNodesArray, &diffNodesArray);
+	}
+
+	/*
+	 * When we're alone in the group, and also when there's no change, then we
+	 * are done here already.
+	 */
+	if (otherNodesArray->count == 0 || diffNodesArray.count == 0)
+	{
+		return true;
+	}
+
+	log_info("Fetched current list of %d other nodes from the monitor "
+			 "to update HBA rules, including %d changes.",
+			 otherNodesArray->count, diffNodesArray.count);
+
+	/*
+	 * We have a new list of other nodes, update the HBA file. We only update
+	 * the nodes that we didn't know before, or that have a new host property.
+	 */
+	if (!keeper_update_group_hba(keeper, &diffNodesArray))
+	{
+		log_error("Failed to update the HBA entries for the new "
+				  "elements in the our formation \"%s\" and group %d",
+				  keeper->config.formation,
+				  keeper->state.current_group);
+		return false;
```

Comment:
If this fails, we won't ever retry adding the HBA entry. Because `keeper->otherNodes` is already modified by `monitor_get_other_nodes`. We should only update keeper->otherNodes  once this succeeded, or we exited early because of `otherNodesArray->count == 0 || diffNodesArray.count == 0`

Reply:
Fixed now by re-installing the “current” version of the array when we fail to update the HBA, and some comments about why it appears safe to proceed that way.